src_features/provideDomainName/cmd_provide_domain_name.c

```diff
@@ -0,0 +1,567 @@
+                    feed_tlv_hash(&data, sig_ctx);
+                }
+                offset += data.length;
+                step = TLV_TAG;
+                break;
+
+            default:
+                apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+                return false;
+        }
+    }
+    // prevent missing or duplicated tags
+    for (int idx = 0; idx < (int) ARRAY_SIZE(handlers); ++idx) {
+        if (handlers[idx].found != 1) {
+            PRINTF("Found %u occurence(s) of tag 0x%x in TLV!\n",
+                   handlers[idx].found,
+                   handlers[idx].tag);
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * Allocate and assign TLV payload
+ *
+ * @param[in] payload payload structure
+ * @param[in] size size of the payload
+ * @return whether it was successful
+ */
+static bool set_payload(s_tlv_payload *payload, uint16_t size) {
+    if ((payload->ptr = mem_alloc(size)) == NULL) {
+        apdu_response_code = APDU_RESPONSE_INSUFFICIENT_MEMORY;
+        return false;
+    }
+    payload->expected_size = size;
+    return true;
+}
+
+/**
+ * Deallocate and unassign TLV payload
+ *
+ * @param[in] payload payload structure
+ */
+static void unset_payload(s_tlv_payload *payload) {
+    mem_dealloc(payload->expected_size);
+    memset(payload, 0, sizeof(*payload));
+}
+
+static bool handle_first_chunk(const uint8_t **data, uint8_t *length, s_tlv_payload *payload) {
+    // check if no payload is already in memory
+    if (payload->ptr != NULL) {
+        unset_payload(payload);
+        apdu_response_code = APDU_RESPONSE_INVALID_P1_P2;
+        return false;
+    }
+
+    // check if we at least get the size
+    if (*length < sizeof(payload->expected_size)) {
+        apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+        return false;
+    }
+    if (!set_payload(payload, U2BE(*data, 0))) {
+        apdu_response_code = APDU_RESPONSE_INSUFFICIENT_MEMORY;
+        return false;
+    }
+
+    // skip the size so we can process it like a following chunk
+    *data += sizeof(payload->expected_size);
+    *length -= sizeof(payload->expected_size);
+    return true;
+}
+
+/**
+ * Handle domain name APDU
+ *
+ * @param[in] p1 first APDU instruction parameter
+ * @param[in] p2 second APDU instruction parameter
+ * @param[in] data APDU payload
+ * @param[in] length payload size
+ */
+void handle_provide_domain_name(uint8_t p1, uint8_t p2, const uint8_t *data, uint8_t length) {
+    s_sig_ctx sig_ctx;
+
+    (void) p2;
+    if (p1 == P1_FIRST_CHUNK) {
+        if (!handle_first_chunk(&data, &length, &g_tlv_payload)) {
+            return response_to_domain_name(false, 0);
+        }
+    } else {
+        // check if a payload is already in memory
+        if (g_tlv_payload.ptr == NULL) {
+            apdu_response_code = APDU_RESPONSE_INVALID_P1_P2;
+            return response_to_domain_name(false, 0);
+        }
+    }
+
+    if (g_tlv_payload.size < g_tlv_payload.expected_size) {
```

Comment:
In which case this condition is false ? I think the second check next line would cover all cases

Reply:
Indeed this was redundant. :+1: